Extend the Reflex class with `now :: PushM t (Event t ())`. (Issue #414)

parenthetical opened the pull request:
I hope the implementation in Spider is correct, it produced the expected result when testing with `switchHoldPromptly never . pushAlways (\a -> fmap (a <$) now)`.

Extending the Reflex class might break modules depending on Reflex if they or another dependency use `now` as an identifier. Let me know if that's not okay.

The change is motivated by distributed Reflex (coming soon™), in which
you can write things like:
```
roundTrip :: (...) => Location t -> m (Event t ())
roundTrip to = headE =<< perceptionsHere =<< perceptionsAt to =<< now
```
Maybe there are other use cases as well?

ryantrinkle commented:
Overall this looks good.  Would you mind adding a new EventSubscribed, instead of reusing `eventSubscribedNever`?  They probably ought to be the same, except for, e.g. https://github.com/reflex-frp/reflex/blob/446094b733aa6d1e51418d0994889d2cc04fca43/src/Reflex/Spider/Internal.hs#L599

Also, we'll need some tests for this.

parenthetical commented:
I'll change that and try to add some tests as well, thanks!

parenthetical commented:
Does this look better?

ryantrinkle commented:
Can we add a test showing that it is *not* firing at some times other than the time it's created?  I think it might need an IORef or something - this looks like it is probably *always* firing (but only the first time each subscriber looks at it).

parenthetical commented:
Thanks for the hints; I implemented it with an IORef and added a test case for the always firing issue.
Was `scheduleClear` was the right procedure to use?

ryantrinkle commented:
Awesome!  That looks right to me :)

Only two things left:
1. You should add something to [the changelog](https://github.com/reflex-frp/reflex/blob/develop/ChangeLog.md) announcing this!
2. We'll take a look at what's going wrong with CI - I think that might be on our end.

parenthetical commented:
Great, thanks for the shepherding!
However, I just tried writing a program with it and now I'm not sure whether adding `now` to `Reflex` was the right thing to do, since I immediately needed a `HasNow` type class as well (which then has to use a different identifier ideally).

ryantrinkle commented:
Hmm, well, putting it in a typeclass seems like a perfectly reasonable thing to do.  Note that not all implementations of MonadSample or MonadHold will have implementations for `now`.  In particular: I think it'd be reasonable for people to assume that `hold a . (b <$) =<< now` will always update the Behavior to `b`, but that's not possible outside of a frame.  And, of course, `PullM` can't use `now` because cached results would be wrong.

3noch commented on src/Reflex/Class.hs:
```diff
@@ -422,6 +420,8 @@ class MonadSample t m => MonadHold t m where
   -- the supplied 'Event'.
   headE :: Event t a -> m (Event t a)
+  -- | An event which occurs at the current moment.
+  now :: m (Event t ())
```
Perhaps it would make sense to add a default signature based on MonadTrans so that all these new methods can be added automatically.

parenthetical replied:
Great idea, I've added it.

3noch commented on src/Reflex/Class.hs:
```diff
@@ -420,6 +420,10 @@ class MonadSample t m => MonadHold t m where
   -- the supplied 'Event'.
   headE :: Event t a -> m (Event t a)
+  -- | An event which occurs at the current moment.
+  now :: m (Event t ())
```
Can any laws be documented?

parenthetical replied:
I added this if that's what you're thinking of?
```
  -- | An event which only occurs at the current moment in time, such that:
  --
  -- > coincidence (pushAlways (\a -> (a <$) <$> now) e) = e
  --
  now :: m (Event t ())
```
Also, `now` turns events in the context of a `MonadHold` into an applicative functor and monad using the existing `Apply` and `Bind` instances:
```
instance (..., MonadHold t m) => Applicative (m . Event t) where
  pure a = (a <$) <$> now
  (<$>) = liftA2 (<.>)
```

ryantrinkle commented:
Awesome, thanks!
